### PR TITLE
Changes loadLibrary -> loadLibraries and readLibrary -> readLibraries.

### PR DESCRIPTION
**Pull request by feldstj**

### CODE REVIEW

**Review by bernardkwok — COMMENTED** [review]
Looks good. 
Can you add a new unit test which has the failure case where the definition is in another library file.
Thanks.

**Comment by niklasharrysson** [comment]
I'm curious about the reason for this change. What was the failure case Bernard mentioned?

It's seams a little bit awkward that a user of the API now has to always supply a vector of paths, even if there is only a single file containing the library to load. To my mind a single file is the most common case, as can be seen in all the updates needed in the test code here.

**Comment by feldstj** [comment]
> I'm curious about the reason for this change. What was the failure case Bernard mentioned?
> 
> It seams a little bit awkward that a user of the API now has to always supply a vector of paths, even if there is only a single file containing the library to load. To my mind a single file is the most common case, as can be seen in all the updates needed in the test code here.

Hey Niklas, the problem was that in RtFileIo::readLibrary(...) the code calls loadLibrary(...) followed by readDocument(...) but since it was being called on one library at a time, if libraryA is dependent on libraryB we throw an exception on readDocument(...) when reading libraryA. The solution is to call loadLibrary(...) on all of the libraries before calling readDocument(...) which fixes the issue.

It's true for the tests I didn't bother changing things. I probably could change things to a single readLibraries(...) call, but I didn't want to change the behavior in case it was necessary the way it was.

**Comment by niklasharrysson** [comment]
> > I'm curious about the reason for this change. What was the failure case Bernard mentioned?
> > It seams a little bit awkward that a user of the API now has to always supply a vector of paths, even if there is only a single file containing the library to load. To my mind a single file is the most common case, as can be seen in all the updates needed in the test code here.
> 
> Hey Niklas, the problem was that in RtFileIo::readLibrary(...) the code calls loadLibrary(...) followed by readDocument(...) but since it was being called on one library at a time, if libraryA is dependent on libraryB we throw an exception on readDocument(...) when reading libraryA. The solution is to call loadLibraries(...) on all of the libraries before calling readDocument(...) which fixes the issue.
> 
> It's true for the tests I didn't bother changing things. I probably could change things to a single loadLibraries(...) call, but I didn't want to change the behavior in case it was necessary the way it was.

Thanks for the clarification and added test case. So this is for the special case when the files that makes up the library to load does not have a common root directory. If they had a common root you could just point to that single root path and all files underneath would be loaded.

Since this is a special case I wonder if we should just add an overloaded function for this, and keep the old signature as is. Because I think the common use-case is to point out a single directory path that holds all the files of a library, as we do with all the standard libraries for example (stdlib, pbrlib, bxdf, etc). It would be unfortunate if all users of this API function would have to construct a vector containing this single path all the time. I'd rather hide this from the API user for convenience. 

So in practice keeping two signatures, where one is just hiding the vector construction:

```
RtStagePtr loadLibrary(const RtString& name, const FilePathVec& paths, const RtReadOptions* options = nullptr, bool forceReload = false);

RtStagePtr loadLibrary(const RtString& name, const FilePath& path, const RtReadOptions* options = nullptr, bool forceReload = false)
{
    return loadLibrary(name, { path }, options, forceReload);
}
```

And I would prefer keeping the name singular `loadLibrary` since it's a single named library being constructed even in the "multiple paths" case. With this change all the test code and other uses of this function can remain unchanged.

**Comment by bernardkwok** [comment]
> > > I'm curious about the reason for this change. What was the failure case Bernard mentioned?
> > > It seams a little bit awkward that a user of the API now has to always supply a vector of paths, even if there is only a single file containing the library to load. To my mind a single file is the most common case, as can be seen in all the updates needed in the test code here.
> > 
> > 
> > Hey Niklas, the problem was that in RtFileIo::readLibrary(...) the code calls loadLibrary(...) followed by readDocument(...) but since it was being called on one library at a time, if libraryA is dependent on libraryB we throw an exception on readDocument(...) when reading libraryA. The solution is to call loadLibraries(...) on all of the libraries before calling readDocument(...) which fixes the issue.
> > It's true for the tests I didn't bother changing things. I probably could change things to a single loadLibraries(...) call, but I didn't want to change the behavior in case it was necessary the way it was.
> 
> Thanks for the clarification and added test case. So this is for the special case when the files that makes up the library to load does not have a common root directory. If they had a common root you could just point to that single root path and all files underneath would be loaded.
> 
> Since this is a special case I wonder if we should just add an overloaded function for this, and keep the old signature as is. Because I think the common use-case is to point out a single directory path that holds all the files of a library, as we do with all the standard libraries for example (stdlib, pbrlib, bxdf, etc). It would be unfortunate if all users of this API function would have to construct a vector containing this single path all the time. I'd rather hide this from the API user for convenience.
> 
> So in practice keeping two signatures, where one is just hiding the vector construction:
> 
> ```
> RtStagePtr loadLibrary(const RtString& name, const FilePathVec& paths, const RtReadOptions* options = nullptr, bool forceReload = false);
> 
> RtStagePtr loadLibrary(const RtString& name, const FilePath& path, const RtReadOptions* options = nullptr, bool forceReload = false)
> {
>     return loadLibrary(name, { path }, options, forceReload);
> }
> ```
> 
> And I would prefer keeping the name singular `loadLibrary` since it's a single named library being constructed even in the "multiple paths" case. With this change all the test code and other uses of this function can remain unchanged.

Well it's not really just this "special case". Any set of libraries can be loaded in any order and if we make the general use case presented to load "one at a time", we can always run into problems. e.g. if you load in bxdf before pbrlib. We happen to have a "special case" for the default libraries which happen to be under "libraries", but it's not guaranteed for any general set of libraries.

Aside from naming which does not match core BTW, I'd suggest using a `FileSearchPath`. basically it's where your going to search for library files.  I can be constructed from a FilePath / string if desired so is backwards compatible.

BTW, the dependency issues doesn't really exist at the core level but at the runtime level as the core->runtime conversion occurs and fails for any missing definitions. The logic (good or bad) in core is you can load in and validation will find out bad dependencies. There is no load librar(ies) in arbitrary order and then convert to runtime logic at this time and this sounds like an extra step that you don't want to expose.

File: resources/Materials/TestSuite/stdlib/definition/within2_color3.mtlx [hunk 1]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0"?>
+<materialx version="1.38">
+  <nodedef name="adsk:ND_within2_color3" node="within2" version="1.0" isdefaultversion="true" nodegroup="BenoitsTest" namespace="adsk">
+    <input name="amplitude" type="vector3" value="1, 1, 1" uivisible="true" />
+    <output name="out" type="color3" value="0, 0, 0" />
+  </nodedef>
+  <nodegraph name="NG_within2_color3" xpos="-19.5049" ypos="739.646" ldx_io_nodes="&amp;lt;?xml version=&quot;1.0&quot;?&amp;gt;&#10;&amp;lt;materialx version=&quot;1.38&quot;&amp;gt;&#10;  &amp;lt;layout name=&quot;io&quot; type=&quot;&quot;&amp;gt;&#10;    &amp;lt;output name=&quot;Output&quot; type=&quot;&quot; ports=&quot;out&quot; pos=&quot;225.333 -61&quot; /&amp;gt;&#10;  &amp;lt;/layout&amp;gt;&#10;&amp;lt;/materialx&amp;gt;&#10;" namespace="adsk" nodedef="adsk:ND_within2_color3">
```
Comment (bernardkwok) [inline resources/Materials/TestSuite/stdlib/definition/within2_color3.mtlx:7]: Can we get rid of the ldx_io_nodes as it's not a standard attribute and isn't required.

File: resources/Materials/TestSuite/stdlib/definition2/TEST_COMPOUND_PUBLISH_color3.mtlx [hunk 2]
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0"?>
+<materialx version="1.38">
+  <nodedef name="ND_TEST_COMPOUND_PUBLISH_color3" node="TEST_COMPOUND_PUBLISH" version="1.0" isdefaultversion="true" nodegroup="BenoitsTest">
+    <input name="amplitude" type="vector3" value="1, 1, 1" />
+    <input name="in2" type="color3" value="1, 1, 1" />
+    <output name="out" type="color3" value="0, 0, 0" />
+  </nodedef>
+  <nodegraph name="NG_TEST_COMPOUND_PUBLISH_color3" xpos="300.624" ypos="-92.5505" ldx_io_nodes="&amp;lt;?xml version=&quot;1.0&quot;?&amp;gt;&#10;&amp;lt;materialx version=&quot;1.38&quot;&amp;gt;&#10;  &amp;lt;layout name=&quot;io&quot; type=&quot;&quot;&amp;gt;&#10;    &amp;lt;input name=&quot;Input&quot; type=&quot;&quot; ports=&quot;amplitude,in2&quot; pos=&quot;-450 -61&quot; /&amp;gt;&#10;  &amp;lt;/layout&amp;gt;&#10;&amp;lt;/materialx&amp;gt;&#10;" nodedef="ND_TEST_COMPOUND_PUBLISH_color3">
```
Comment (bernardkwok) [inline resources/Materials/TestSuite/stdlib/definition2/TEST_COMPOUND_PUBLISH_color3.mtlx:8]: Smae ldx_io_nodes comment. Please remove.

File: source/MaterialXTest/MaterialXRuntime/Runtime.cpp [hunk 3]
```diff
@@ -2549,10 +2549,10 @@ TEST_CASE("Runtime: graph output connection", "[runtime]")
     mx::FileSearchPath searchPath(mx::FilePath::getCurrentPath() /
                                   mx::FilePath("libraries"));
     api->setSearchPath(searchPath);
-    api->loadLibrary(TARGETS_NAME, RuntimeGlobals::TARGETS_PATH());
-    api->loadLibrary(STDLIB_NAME, RuntimeGlobals::STDLIB_PATH());
-    api->loadLibrary(PBRLIB_NAME, RuntimeGlobals::PBRLIB_PATH());
-    api->loadLibrary(BXDFLIB_NAME, RuntimeGlobals::BXDFLIB_PATH());
+    api->loadLibraries(TARGETS_NAME, { RuntimeGlobals::TARGETS_PATH() });
```
Comment (bernardkwok) [inline source/MaterialXTest/MaterialXRuntime/Runtime.cpp:2552]: Would be good to change this to use a single call to "promote" that interface for this case as it's really not "safe" to load one at a time, especially if the order is incorrect.

**Comment by niklasharrysson** [comment]
> Well it's not really just this "special case". Any set of libraries can be loaded in any order and if we make the general use case presented to load "one at a time", we can always run into problems. e.g. if you load in bxdf before pbrlib. We happen to have a "special case" for the default libraries which happen to be under "libraries", but it's not guaranteed for any general set of libraries.
> 
> Aside from naming which does not match core BTW, I'd suggest using a `FileSearchPath`. basically it's where your going to search for library files. It can be constructed from a FilePath / string if desired so is backwards compatible.
> 
> BTW, the dependency issues doesn't really exist at the core level but at the runtime level as the core->runtime conversion occurs and fails for any missing definitions. The logic (good or bad) in core is you can load in and validation will find out bad dependencies. There is no load librar(ies) in arbitrary order and then convert to runtime logic at this time and this sounds like an extra step that you don't want to expose.

Aha, so now I understand the problem better. You want to load in a set of different libraries in one batch. That is not what this original function was intended to do. It was meant to load a single library, assigning it a name to identify it, to be able to reload it later if things have changed or unload it if not needed any more. So more fine grained library management, something you can't do in core at all. As you have noted this comes with the requirement that you have to load the libraries with knowledge of their dependencies, and in right order.

If you load many things in one batch you don't have to know the _order_, but you still have to know about dependencies and make sure every dependency is represented in the vector of paths given. If we have two functions, as above, the client can choose which approach is better for their use case.

Either way we're not really solving the document dependency issues we have in MaterialX. It would be nice if we had a way to specify dependencies in the syntax, other than the xinclude statement that just dumps everything into a single core document. But I guess we're getting out of scope for this PR now :)

**Comment by niklasharrysson** [comment]
> Aside from naming which does not match core BTW, I'd suggest using a `FileSearchPath`. basically it's where your going to search for library files. It can be constructed from a FilePath / string if desired so is backwards compatible.

A search path is already used, as the doc string say if the given path is relative the search path set with `RtApi::setSearchPath` will be used.
```
/// Load a file or folder of files into a new library. The search paths previously set
/// will be used to find the files if the given path is a relative path.
```

**Comment by bernardkwok** [comment]
> > Well it's not really just this "special case". Any set of libraries can be loaded in any order and if we make the general use case presented to load "one at a time", we can always run into problems. e.g. if you load in bxdf before pbrlib. We happen to have a "special case" for the default libraries which happen to be under "libraries", but it's not guaranteed for any general set of libraries.
> > Aside from naming which does not match core BTW, I'd suggest using a `FileSearchPath`. basically it's where your going to search for library files. It can be constructed from a FilePath / string if desired so is backwards compatible.
> > BTW, the dependency issues doesn't really exist at the core level but at the runtime level as the core->runtime conversion occurs and fails for any missing definitions. The logic (good or bad) in core is you can load in and validation will find out bad dependencies. There is no load librar(ies) in arbitrary order and then convert to runtime logic at this time and this sounds like an extra step that you don't want to expose.
> 
> Aha, so now I understand the problem better. You want to load in a set of different libraries in one batch. That is not what this original function was intended to do. It was meant to load a single library, assigning it a name to identify it, to be able to reload it later if things have changed or unload it if not needed any more. So more fine grained library management, something you can't do in core at all. As you have noted this comes with the requirement that you have to load the libraries with knowledge of their dependencies, and in right order.
> 
> If you load many things in one batch you don't have to know the _order_, but you still have to know about dependencies and make sure every dependency is represented in the vector of paths given. If we have two functions, as above, the client can choose which approach is better for their use case.
> 
> Either way we're not really solving the document dependency issues we have in MaterialX. It would be nice if we had a way to specify dependencies in the syntax, other than the xinclude statement that just dumps everything into a single core document. But I guess we're getting out of scope for this PR now :)

Your correct. There are still many ways to "do it incorrectly" :) as there are no dependency relationships. There was at some point the idea of having a document "reference" another document but that was started and then halted. Some concept like a module which could collect related definitions and module-to-module dependency might be a usef granularity -- but as you say a bit out of scope.

If we go with two interfaces then, all that I'd like to see is that it's clear that one interface is for the "single" case, and the new interface is for a "collection" case. Though I still believe a single interface with a FileSearchPath can work. The idea being that the "name" of the library does not need to match the files being loaded in but more a semantic name. How does that sound ? So
stdlib et al are represent the "core" which is compose of files in N paths.

File: source/MaterialXTest/MaterialXRuntime/Runtime.cpp [hunk 4]
```diff
@@ -1472,7 +1469,7 @@ TEST_CASE("Runtime: Traversal", "[runtime]")
 {
     mx::RtScopedApiHandle api;
 
-    mx::FileSearchPath searchPath(mx::FilePath::getCurrentPath() / mx::FilePath("libraries"));
+    mx::FileSearchPath searchPath(mx::FilePath::getCurrentPath() / "libraries");
```
Comment (bernardkwok) [inline source/MaterialXTest/MaterialXRuntime/Runtime.cpp:1472]: Guess we should also update this one and the one at line 1919 ?

**Review by bernardkwok — APPROVED** [review]
Looks good. Thanks.

File: resources/Materials/TestSuite/stdlib/definition/within2_color3.mtlx [hunk 5]
```diff
@@ -0,0 +1,14 @@
+<?xml version="1.0"?>
+<materialx version="1.38">
+  <nodedef name="adsk:ND_within2_color3" node="within2" version="1.0" isdefaultversion="true" nodegroup="BenoitsTest" namespace="adsk">
```
Comment (bernardkwok) [inline resources/Materials/TestSuite/stdlib/definition/within2_color3.mtlx:3]: Sorry I missed this, can we remove Benosts name and the adsk namespace for both files :).

**Pull request merged**